File: src/06-B-ademe_merge.R
library(tidyverse)
library(data.table)
library(arrow)

TABLE_PASSAGE_DF <- read_feather("results_building/t_passage.feather") |>
  select(code_insee24, arr24, bv2022, code_insee21)

ARR2COM <- read_feather("results_building/arr2com.feather") 

old_replaced <- open_dataset("data/interim/_ademe/dpe_all.feather", format = "arrow") |> 
  select(ndperemplac) |> 
  unique() |>
  collect() |>
  filter(ndperemplac!="")

old_building <- open_dataset("data/interim/_ademe/dpe_all.feather", format = "arrow") |> 
  select(ndpeimmeubleassoci) |>
  unique() |>
  collect() |>
  filter(ndpeimmeubleassoci!="")

old_filter <- data.table(dpe_id = c(old_replaced[[1]], old_building[[1]]), filt = T)

OLD_DPE <- open_dataset("data/interim/_ademe/dpe_all.feather", format = "arrow") |> 
  filter(modledpe=="DPE 3CL 2021 méthode logement")|>
  select(
    dpe_id = ndpe,
    dpe_replaced = ndperemplac,
    dpe_building = ndpeimmeubleassoci,
    date_visited = datevisitediagnostiqueur, 
    date_established = datetablissementdpe,
    type = modledpe,
    floor_area = surfacehabitablelogement, 
    ac_area = surfaceclimatise,
    ghg_grade = etiquetteges,
    dpe_grade = etiquettedpe, 
    year_built = anneconstruction, 
    period_built = priodeconstruction,
    insee_code = codeinseeban,
    heating_energy = typenergieprincipalechauffage,
    heating_type = typeinstallationchauffage
  ) |>
  mutate("insee_code" = str_remove_all(str_remove(insee_code, "old"), " ")) |>
  left_join(ARR2COM, by = c("insee_code"="code_arr")) |>
  mutate(insee_code = if_else(is.na(code_com), insee_code, code_com)) |>
  left_join(TABLE_PASSAGE_DF, by=c("insee_code"="code_insee21")) |>
  collect() |>
  setDT()

OLD_DPE <- old_filter[OLD_DPE, on = .(dpe_id)]
OLD_DPE <- OLD_DPE[is.na(filt)][, "filt" := NULL]

OLD_DPE[, "period" := fcase(
  period_built=="1948-1974", 1961,
  period_built=="1975-1977", 1976,
  period_built=="1978-1982", 1980, 
  period_built=="1983-1988", 1987,
  period_built=="1989-2000", 1995, 
  period_built=="2001-2005", 2003, 
  period_built=="2006-2012", 2009, 
  period_built=="2013-2021", 2016, 
  period_built=="après 2021", 2022, 
  period_built=="avant 1948", 1935
)]

OLD_DPE[, "ac_area" := if_else(is.na(ac_area), 0, ac_area)]
OLD_DPE[, "year_built" := if_else(is.na(year_built), period, year_built)]

DPE_ENCODER <- data.table(
  dpe_grade = c("A", "B", "C", "D", "E", "F", "G"),
  dpe_grade_num = 1:7
)

OLD_DPE <- OLD_DPE[DPE_ENCODER, on = .(dpe_grade)]

group_dpe <- function(OLD_DPE, groupvar){
  
  DPE_DF <- OLD_DPE[, 
    .(built_q2 = median(year_built, na.rm=T), floor_area_q2 = median(floor_area, na.rm=T),
      q1_dpe = quantile(dpe_grade_num, .25), q2_dpe = quantile(dpe_grade_num, .5), 
      q3_dpe = quantile(dpe_grade_num, .75), n_dpe = .N), 
    by = groupvar
  ]

  AC_DF <- OLD_DPE[ac_area >0, 
    .(ac_prop = .N), 
    by= groupvar
  ]

  DPE_DF <- DPE_DF |>
  mutate(iqr_dpe_num = round(q3_dpe) - round(q1_dpe)) |>
  pivot_longer(c(q1_dpe, q2_dpe, q3_dpe)) |>
  mutate(value = round(value, digits = 0)) |>
  pivot_wider(names_from = name, values_from = value) |>
  left_join(AC_DF, {{groupvar}}) |>
  mutate(ac_prop = ac_prop/n_dpe)

  return(DPE_DF)

}

OLD_DPE_ARR <- group_dpe(OLD_DPE, "arr24") |>
  select(arr24, everything()) |>
  filter(nchar(arr24)<4) |>
  arrange(arr24)

OLD_DPE_BV <- group_dpe(OLD_DPE, "bv2022") |>
  select(bv2022, everything()) |>
  filter(!is.na(bv2022)) |>
  arrange(bv2022)

OLD_DPE_DEP <- OLD_DPE |>
  mutate(dep = substr(arr24, 1, 2)) |>
  group_dpe("dep") |>
  select(dep, everything()) |>
  filter(!is.na(dep)) |>
  arrange(dep)

write_feather(as.data.frame(OLD_DPE_ARR), "data/interim/arrondissements/ademe_dpe.feather", compression = "zstd")
write_feather(as.data.frame(OLD_DPE_BV), "data/interim/bv2022/ademe_dpe.feather", compression = "zstd")
write_feather(as.data.frame(OLD_DPE_DEP), "data/interim/departement/ademe_dpe.feather", compression = "zstd")

File: src/06-B-ademe_merge.py
import polars as pl
import pyarrow.feather as feather
from pathlib import Path


# Read support_tables
table_passage = (
    pl.read_ipc("results_building/t_passage.feather")
    .select(["code_insee24", "arr24", "bv2022", "code_insee21"])
)
arr2com = pl.read_ipc("results_building/arr2com.feather")

# full DPE dataset
dpe_all = pl.read_ipc("data/interim/_ademe/dpe_all.feather")

# building-associated IDs
old_replaced = (
    dpe_all
    .select("ndperemplac")
    .unique()
    .filter(pl.col("ndperemplac") != "")
    .to_series()
    .to_list()
)
old_building = (
    dpe_all
    .select("ndpeimmeubleassoci")
    .unique()
    .filter(pl.col("ndpeimmeubleassoci") != "")
    .to_series()
    .to_list()
)
old_filter = pl.DataFrame({
    "dpe_id": old_replaced + old_building,
    "filt": True
})

# Preprocess DPE data
dpe = (
    dpe_all
    .filter(pl.col("modledpe") == "DPE 3CL 2021 méthode logement")
    .select([
        pl.col("ndpe").alias("dpe_id"),
        pl.col("ndperemplac").alias("dpe_replaced"),
        pl.col("ndpeimmeubleassoci").alias("dpe_building"),
        pl.col("datevisitediagnostiqueur").alias("date_visited"),
        pl.col("datetablissementdpe").alias("date_established"),
        pl.col("modledpe").alias("type"),
        pl.col("surfacehabitablelogement").alias("floor_area"),
        pl.col("surfaceclimatise").alias("ac_area"),
        pl.col("etiquetteges").alias("ghg_grade"),
        pl.col("etiquettedpe").alias("dpe_grade"),
        pl.col("anneconstruction").alias("year_built"),
        pl.col("priodeconstruction").alias("period_built"),
        pl.col("codeinseeban").alias("insee_code"),
        pl.col("typenergieprincipalechauffage").alias("heating_energy"),
        pl.col("typeinstallationchauffage").alias("heating_type"),
    ])
    .with_columns([
        pl.col("insee_code").str.replace_all("old", "").str.replace_all(" ", "").alias("insee_code")
    ])
    .join(arr2com, left_on="insee_code", right_on="code_arr", how="left")
    .with_columns([
        pl.when(pl.col("code_com").is_null())
        .then(pl.col("insee_code"))
        .otherwise(pl.col("code_com"))
        .alias("insee_code")
    ])
    .drop(["code_com"])
    .join(table_passage, left_on="insee_code", right_on="code_insee21", how="left")
)

# Exclude old entries
dpe = (
    dpe
    .join(old_filter, on="dpe_id", how="left")
    .filter(pl.col("filt").is_null())
    .drop("filt")
)

# Create mapping tables
period_map_df = pl.DataFrame({
    "period_built": [
        "1948-1974", "1975-1977", "1978-1982", "1983-1988",
        "1989-2000", "2001-2005", "2006-2012", "2013-2021",
        "après 2021", "avant 1948"
    ],
    "period": [1961, 1976, 1980, 1987, 1995, 2003, 2009, 2016, 2022, 1935]
})
grade_map_df = pl.DataFrame({
    "dpe_grade": list("ABCDEFG"),
    "dpe_grade_num": list(range(1, 8))
})

# Map construction periods and grades via joins
dpe = (
    dpe
    .join(period_map_df, on="period_built", how="left")
    .join(grade_map_df, on="dpe_grade", how="left")
)

# Period building
dpe = (
    dpe
    .with_columns([
        pl.col("ac_area").fill_null(0),
        pl.when(pl.col("year_built").is_null())
            .then(pl.col("period"))
            .otherwise(pl.col("year_built"))
            .alias("year_built")
    ])
)

# Define grouping function
def group_dpe(df: pl.DataFrame, groupvar: str) -> pl.DataFrame:
    agg = df.group_by(groupvar).agg([
        pl.col("year_built").median().alias("built_q2"),
        pl.col("floor_area").median().alias("floor_area_q2"),
        pl.col("dpe_grade_num").quantile(0.25).round().alias("q1_dpe"),
        pl.col("dpe_grade_num").quantile(0.5).round().alias("q2_dpe"),
        pl.col("dpe_grade_num").quantile(0.75).round().alias("q3_dpe"),
        pl.len().alias("n_dpe"),
    ])
    ac = df.filter(pl.col("ac_area") > 0).group_by(groupvar).agg([
        pl.len().alias("ac_count")
    ])
    res = (
        agg.join(ac, on=groupvar, how="left")
            .with_columns([
                (pl.col("ac_count") / pl.col("n_dpe")).alias("ac_prop"),
                (pl.col("q3_dpe") - pl.col("q1_dpe")).alias("iqr_dpe_num")
            ])
    ).drop("ac_count")
    return res

# Generate summaries
OLD_DPE_ARR = (
    group_dpe(dpe, "arr24")
    .filter(pl.col("arr24").str.len_chars() < 4)
    .sort("arr24")
)
OLD_DPE_BV = (
    group_dpe(dpe, "bv2022")
    .filter(pl.col("bv2022").is_not_null())
    .sort("bv2022")
)
dep_df = dpe.with_columns([pl.col("arr24").str.slice(0, 2).alias("dep")])
OLD_DPE_DEP = (
    group_dpe(dep_df, "dep")
    .filter(pl.col("dep").is_not_null())
    .sort("dep")
)

# Export files
for df_out, subdir in [
    (OLD_DPE_ARR, "arrondissements"),
    (OLD_DPE_BV, "bv2022"),
    (OLD_DPE_DEP, "departement"),
]:
    out_dir = Path(f"data/interim/{subdir}")
    out_dir.mkdir(parents=True, exist_ok=True)
    feather.write_feather(df_out.to_arrow(), out_dir / "ademe_dpe.feather", compression="zstd")


